src/journal-remote/journal-gatewayd.c

```diff
@@ -885,6 +885,66 @@ static int request_handler_machine(
         return MHD_queue_response(connection, MHD_HTTP_OK, response);
 }
 
+static int request_handler_boots(
+                struct MHD_Connection *connection,
+                void *connection_cls) {
+
+        _cleanup_(MHD_destroy_responsep) struct MHD_Response *response = NULL;
+        RequestMeta *m = ASSERT_PTR(connection_cls);
+        int r;
+        _cleanup_free_ LogId *boots = NULL;
+        size_t n_boots;
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *array = NULL;
+        _cleanup_free_ char *json = NULL;
+
+        assert(connection);
+
+        r = open_journal(m);
+        if (r < 0)
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to open journal: %m");
+
+        r = journal_get_boots(
+                        m->journal,
+                        false,
+                        SIZE_MAX,
+                        &boots, &n_boots);
+
+        if (r < 0)
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to determine boots: %m");
+        if (r == 0)
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "No boot found.");
+
+        for (int i = 0; i < (int) n_boots; i++) {
+                char boot_id[SD_ID128_STRING_MAX];
+                sd_id128_to_string(boots[i].id, boot_id);
+
+                r = sd_json_variant_append_arrayb(
+                                &array,
+                                SD_JSON_BUILD_OBJECT(
+                                                SD_JSON_BUILD_PAIR("index", SD_JSON_BUILD_INTEGER(-i)),
+                                                SD_JSON_BUILD_PAIR("boot_id", SD_JSON_BUILD_STRING(boot_id)),
+                                                SD_JSON_BUILD_PAIR("first_entry", SD_JSON_BUILD_UNSIGNED(boots[i].first_usec)),
+                                                SD_JSON_BUILD_PAIR("last_entry", SD_JSON_BUILD_UNSIGNED(boots[i].last_usec))));
+                if (r < 0)
+                        return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to append JSON object: %m");
+        }
+
+        r = sd_json_variant_format(array, 0, &json);
+        if (r < 0) {
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to format JSON data: %m");
+        }
```

Comment:
coding style, no {} around single-line if blocks

Reply:
hmm, the number of boots returned might be massive, hence i think it might be wiser to return something streamable (i.e. json-seq here, RFC 7464) rather than a  single huge array. i.e. output each boot individually and use the SD_JSON_FORMAT_SEQ flag.